Add clarification regarding default values

madelson opened the pull request:
Clarify the compatibility implications of changing or removing parameter default values. 

Fix #28200

## Summary

Describe your changes here.

Fixes #Issue_Number (if available)

madelson committed:
Add clarification regarding default values

Clarify the compatibility implications of changing or removing parameter default values. 

Fix #28200

madelson committed:
Attempt to fix linting errors

gewarren commented on docs/core/compatibility/index.md:
```diff
@@ -220,6 +220,10 @@ Changes in this category modify the public surface area of a type. Most of the c
 
 - ❌ **DISALLOWED: Changing the default value of a property, field, or parameter**
 
+  Note that changing or removing a *parameter* default value is not a binary break. Removing a parameter default value is a source break and changing a parameter default value could result in a behavioral break after recompilation.
```
```suggestion
  Changing or removing a *parameter* default value is not a binary break. Removing a parameter default value is a source break, and changing a parameter default value could result in a behavioral break after recompilation.
```

gewarren commented on docs/core/compatibility/index.md:
```diff
@@ -220,6 +220,10 @@ Changes in this category modify the public surface area of a type. Most of the c
 
+  Note that changing or removing a *parameter* default value is not a binary break. Removing a parameter default value is a source break and changing a parameter default value could result in a behavioral break after recompilation.
+
+  For this reason, removing parameter default values is considered acceptible from a compatibility perspective in the specific case of "moving" those default values to a new method overload in order to eliminate ambiguity. For example, if we have a method `Foo(int a = 1)` and we introduce an overload of `Foo` with 2 optional parameters `a` and `b`, we can preserve compatibility my moving `a`'s de ...
```
```suggestion
  For this reason, removing parameter default values is acceptable in the specific case of "moving" those default values to a new method overload to eliminate ambiguity. For example, consider an existing method `MyMethod(int a = 1)`. If you introduce an overload of `MyMethod` with two optional parameters `a` and `b`, you can preserve compatibility by moving the default value of `a` to the new overload. Now the two overloads are `MyMethod(int a)` and `MyMethod(int a = 1, int b = 2)`. This pattern allows `MyMethod()` to compile.
```

gewarren replied:
We're discouraged from using "Foo" as a name.

bartonjs reviewed (approved):
With @gewarren 's suggestions, LGTM

madelson committed:
Update docs/core/compatibility/index.md

Incorporate feedback from https://github.com/dotnet/docs/pull/28224

Co-authored-by: Genevieve Warren <24882762+gewarren@users.noreply.github.com>

madelson committed:
Update docs/core/compatibility/index.md

Incorporate feedback from https://github.com/dotnet/docs/pull/28224#discussion_r806226047

Co-authored-by: Genevieve Warren <24882762+gewarren@users.noreply.github.com>
